lib/gemstash/storage_services/s3_storage.rb

```diff
@@ -0,0 +1,289 @@
+# frozen_string_literal: true
+
+require "gemstash"
+require "digest"
+require "fileutils"
+require "yaml"
+require "aws-sdk-s3"
+require "pathname"
+
+module Gemstash
+  # The entry point into the storage engine for storing cached gems, specs, and
+  # private gems using Amazon S3 service.
+  class S3
+    extend Gemstash::Env::Helper
+    include Gemstash::Env::Helper
+    attr_reader :folder, :client, :bucket
+
+    STORAGE_VERSION = 1
+
+    # If the storage engine detects the base cache directory was originally
+    # initialized with a newer version, this error is thrown.
+    class VersionTooNewError < StandardError
+      def initialize(folder, version)
+        super("Gemstash storage version #{Gemstash::S3::STORAGE_VERSION} does " \
+               "not support version #{version} found at #{folder}")
+      end
+    end
+
+    def initialize(folder, root: true)
+      @folder = folder
+      check_storage_version if root
+      @client = Aws::S3::Client.new(
+        access_key_id: gemstash_env.config[:aws_access_key_id],
+        secret_access_key: gemstash_env.config[:aws_secret_access_key],
+        region: gemstash_env.config[:region]
+      )
+      @bucket_name = gemstash_env.config[:bucket_name]
+      @object_name = @folder
+    end
+
+    def check_credentials?
+      @client.get_bucket_location(bucket: @bucket_name).location_constraint == gemstash_env.config[:region]
+    end
+
+    def delete_with_prefix(prefix = @folder)
+      Aws::S3::Resource.new(client: @client).bucket(@bucket_name).objects(prefix: prefix).batch_delete!
+    end
```

Comment:
@bronzdoc This is the method I was talking about. It is used in https://github.com/duyquangnguyenhac/gemstash/blob/8d51e62440c5b5756971bbc8507ada54c5904e42/spec/gemstash/s3_storage_spec.rb#L23
It is not wise to have such delete method, but I use it to clear the storage after I run the VCR tests. How should I fix this ?

Reply:
You could add this method to Gemstash::S3 only inside the tests. In one of the support files loaded by the tests, something like this:

```ruby
module Gemstash
  class S3
    def delete_with_prefix(prefix = @folder)
       Aws::S3::Resource.new(client: @client).bucket(@bucket_name).objects(prefix: prefix).batch_delete!
     end
  end
end
```

Reply:
Thank you so much for your suggestion. I did move that method into a helper method that's only used during tests!
It is fixed in this commit: e7b7336